### PR TITLE
Add relay service listeners and handlers

### LINKED ISSUE #26: Childchain relayer service
# Bounty
Listener and handler functionalities to relay transactions in the child-chains

## Scope

-  Handler function that receives lock events from CKB relayer service
-  Collect validator signatures via contract method collectLock
-  Listen for burn/unlock events on childchain
-  Once a burn occurs trigger validator signing
-  Collect the validator signature via contract method collectUnlock
-  Listen for the BurnQuorum event
-  Send BurnQuorum  event data to CKB relayer service
-  Use same json file to configure environment for both relayer services
> listener sleep time
> url endpoints for child chain
> url endpoints for ckb relayer service
> current chain mainent/testnet
> bridge script hash
> etc

## Deliverables

- Deploy-able code with instructions on running service
- Service allow's validator to sign transactions for collectUnlock
- Service allow's relay of BurnQuorum to parent bridge

## Funding Circle

Nervos

## Bounty Owner/Gardener

@mwaa as LeadLink

## Roles
bounty gardener: @mwaa / 50 DAI @johannbarbie / 50 DAI 
bounty worker: @mwaa / 700 DAI
bounty reviewer: name / 100 DAI

## Gardener checklist

- [x] I have attached one of the `size-` labels to this bounty
- [ ] I have submitted this bounty via http://bounty.leapdao.org/viewform
- [ ] I have added this bounty to bounty board here https://github.com/orgs/leapdao/projects/6

### LINKED ISSUE #40: CKB relayer service
# Bounty
Listener and handler functionalities to relay transactions for parent bridge

## Scope

-  Handler function that receives unlock events from child-chain relay service
-  Parse event data and call necessary transaction helpers to construct transaction that uses payout script
-  Use ckb-indexer and listen on bridge related transactions
-  Listen for lock events from parent bridge. i.e filter out transactions based on deposit script hash
-  Send event data to child-chain relay handler function
-  Use same json file to configure environment for both relayer services
> listener sleep time
> url endpoints for child chain
> url endpoints for ckb relayer service
> current chain mainent/testnet
> bridge script hash
> etc

## Deliverables

- Deploy-able code with instructions on running service
- Service sends lock event to child-chain relayer service
- Service handler recieves unlock event from child-chain relayer service

## Funding Circle

Nervos

## Bounty Owner/Gardener

@mwaa as LeadLink

## Roles
bounty gardener: @mwaa / 100 DAI 
bounty worker: @mwaa / 370 DAI
bounty reviewer: name / 80 DAI

## Gardener checklist

- [x] I have attached one of the `size-` labels to this bounty
- [ ] I have submitted this bounty via http://bounty.leapdao.org/viewform
- [ ] I have added this bounty to bounty board here https://github.com/orgs/leapdao/projects/6

### PR DESCRIPTION
**Pull request by mwaa**
### How to test

1. Start ethermint
2. Start Redis on port 6379
3. In the relayer folder add childchain bridge address.
4. If not deployed you can run `ts-node deploy.ts`. Copy value of address from address.json to key address in config.json
5. Run `ts-node start.ts`. 
6. Test Unlock by running `ts-node /src/utils/transaction.ts`. Will create 3 separate burn events that should be captured and recorded

Closes #26 Closes #40

### CODE REVIEW

File: packages/relayer/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Bridge
+
+Functionality to listen for events and relay them across chains
+
+### Redis
+
+We use redis two track two things
```
Comment (Clearwood) [inline packages/relayer/README.md:7]:
```suggestion
We use redis to track two things
```

File: packages/relayer/src/evm.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+Contract.setProvider('ws://localhost:8546');
+import Config from '../config.json';
+
+class EVMRelay {
+    db: any;
+    queueRunner: any;
+    contractInstance: any;
+    validatorAddress: string;
+    bridgeAddress: string = Config.address;
+    bridgeHash: string = Config.brigdgeHash;
```
Comment (Clearwood) [inline packages/relayer/src/evm.ts:13]: spelling error in bridgeHash

File: packages/relayer/start.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      set: promisify(redisClient.set).bind(redisClient),
+      quit: promisify(redisClient.quit).bind(redisClient)
+    };
+  
+    return db;
+}
+  
+const startQueue = async () => {
+  const qname = Config.address;
+  console.log("This is the quee", qname);
```
Comment (Clearwood) [inline packages/relayer/start.ts:33]:
```suggestion
  console.log("This is the queue", qname);
```

File: packages/relayer/README.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Bridge
+
+Functionality to listen for events and relay them across chains
+
+### Redis
+
```
Comment (Clearwood) [inline packages/relayer/README.md:6]: Please add instructions on Redis version to use, as well as configuration like the port here. 

A longer description of the Relayer service is missing here, all instructions on how to run the service should be able to be found in this file.

**Comment by Clearwood** [comment]
image
I am getting errors on connecting to Redis even though my instance is started on port 6379

**Comment by mwaa** [comment]
Does not use current childchain start script
1. Clone latest ethermint `git clone git@github.com:ChainSafe/ethermint.git`
2. Run `/.init.sh`
3. On new console listen on web socket port. `ethermintcli rest-server --laddr "tcp://localhost:8545" --wsport 8546 --unlock-key mykey --chain-id ethermint-1 --trace`
4. On relayer folder install packages using `yarn`
5. Deploy contract using `npm run deploy`. Copy address to config file
6. Use `npm run start` to start the service
7. Add console.log on `_collectUnLock` on the `evm.ts` file to confirm receipt object is received
8. Use `npm run dummy:unlock` to trigger 3 dummy burn events
9. Confirm receipt objects are logged on console

**Comment by johannbarbie** [comment]
i've executed all steps as described above. I see the script producing transactions:
```
johba$ npx ts-node src/utils/transaction.ts 
We have transaction 0xe7476ce751935ebabddd9f73b2ae99033c781b34730820adfeb88b8c67bd6401
We have transaction 0x3ed39d55eae12d118878f1e9f2197f72a08a95d87c57026ffbbd5a255c4d6176
We have transaction 0xc46b1f9ee904314fab02def7f19b12c2ba7a1075e22eb92a7ab768779b87267d
```
but console.log is never triggered:
```
$ npx ts-node start.ts
DB has been connected...
No matching redis queue found. Creating a new one.
Queue successfully created...
Listening on childchain...
Blocks to process are 27
Listening on childchain...
Blocks to process are 10
Listening on childchain...
```

**Comment by johannbarbie** [comment]
getting negative block numbers suddenly. stopped everything and did a redeployment, still same.
```
a1708-4:relayer johba$ yarn start
yarn run v1.22.10
warning package.json: No license field
$ npx ts-node start.ts
DB has been connected...
Listening on childchain...
Blocks to process are -92
Listening on childchain...
Blocks to process are -84
Listening on childchain...
Blocks to process are -76
Listening on childchain...
Blocks to process are -68
Listening on childchain...
Blocks to process are -60
Listening on childchain...
Blocks to process are -52
```

**Comment by mwaa** [comment]
Negative block numbers is because we don't reset the redis db when service stops
So if it recorded up to height 400 and you start a new network blocks to process will be negative

Forked out the bug since its taking long #59

**Review by johannbarbie — APPROVED** [review]
code looks good. events forked out and need to be solved.

**Pull request merged**